testing:  increase installer coverage

tldahlgren opened the pull request:
This PR is dedicated to increasing coverage of the new `installer.py` module.

tldahlgren commented:
Note `d5d602f` is just to determine if `coverage combine` is needed to cover the forked `build_process` in `installer._install_task`, which appears to have no coverage despite the fork call being made.

tldahlgren commented:
Reset status to see if that will re-run the tests.

tldahlgren commented:
Interesting.  Looks like explicitly adding `coverage combine` to `run-unit-tests` causes a fatal error:

```
No data to combine
The command "share/spack/qa/run-$TEST_SUITE-tests" exited with 1.
```

So reverted.

tldahlgren commented:
FYI.  As of commit `58a641e`, `installer.py` has 90.61% coverage (i.e., everything but the forked `bulid_process`.

tldahlgren commented:
Very strange.  Coverage for `eeceb04` at https://codecov.io/gh/spack/spack shows counts on comment lines and no coverage for tests that are explicitly passing (e.g., `test_install_task_stop_iter`).

tldahlgren commented:
Sub-Process Coverage Experiment:
Although existing test cases utilize the installer's `build_process`, coverage for those calls are not being picked up.  This loss cannot be attributed to `.coveragerc`, which has the necessary `parallel` and `concurrency` options.

According to the latest `coverage` documentation (https://coverage.readthedocs.io/en/latest/subprocess.html as of March 2020), two things are needed to measure sub-process coverage:

1. a suitably defined COVERAGE_PROCESS_START environment variable and
2. Python configured to invoke `coverage.process_startup()`

The environment variable attempted here was inspired by the demo example at https://pypi.org/project/coverage_enable_subprocess/.  The configuration choice involved a Spack-specific `sitecustomize.py` since it is cleaner than the alternative and can be placed in any directory on the Python path.  The down side is that there can be only one `sitecustomize.py` file so there may still be issues with this approach.  The alternative configuration option is to use a `.pth` file but it must be created in a system-defined and, therefore, privileged directory.

Unfortunately, it is unclear whether these changes cause conflicts with the build environment or affect the results reporting coverage counts of comments and blanks, as examples, described in a previous comment above.

tldahlgren commented:
Ugh!  The test that covers the one line being highlighted as uncovered in `codecov/project` was moved to #15339 .

alalazo commented:
@tldahlgren See #15354

tldahlgren commented:
Strange. One of the merges should be covering the supposedly uncovered 4% reported in the patch coverage.

tldahlgren commented:
> @tldahlgren See #15354

That was great!  Thanks!

alalazo commented on lib/spack/spack/test/installer.py:
```diff
@@ -162,6 +166,25 @@ def _spec(spec):
 
+def test_try_install_from_binary_cache(install_mockery, mock_packages,
+                                       monkeypatch, capsys):
+    """Tests to add coverage to _try_install_from_binary_cache."""
```
As a minor comment, I would prefer the docstrings (if there's need for any) to reflect the behavior they are testing rather than the function they are adding coverage to. Here, for instance, I feel like the name of the test function is more informative than the docstring.

tldahlgren replied:
Excellent point.  I was getting sloppy.

tldahlgren replied:
Fair enough.  I've gone through and modified or removed this and similar docstrings throughout this module.

I've also split test functions when they contained multiple, non-dependent checks.

tldahlgren commented on lib/spack/spack/installer.py:
```diff
@@ -1154,7 +1153,7 @@ def build_process():
             # A StopIteration exception means that do_install was asked to
             # stop early from clients.
-            tty.msg('{0} {1}'.format(self.pid, str(e)))
+            tty.msg('{0} {1}'.format(self.pid, e.message))
```
It is.  Thanks for catching that.

tldahlgren commented on lib/spack/spack/test/install.py:
```diff
@@ -437,32 +437,68 @@ def test_pkg_install_paths(install_mockery):
 
 
-def test_pkg_install_log(install_mockery):
+def test_pkg_install_log_no_log(install_mockery, monkeypatch):
```
I found it handy to test all the log-related tests with `-k test_pkg_install_log` but this name is definitely more clear.  Thanks.

tldahlgren replied:
> Do we need `monkeypatch` here?

No.  Thanks!

tldahlgren replied:
> Also, I would rename the test function something like:

I ended up renaming this to something else after reading another comment you made below.  Hopefully the new name is clearer.

tldahlgren commented on lib/spack/spack/test/install.py:
```diff
@@ -437,32 +437,68 @@ def test_pkg_install_paths(install_mockery):
-    # Set up mock build files and try again
+
+def test_pkg_install_log_with_log(install_mockery, monkeypatch):
+    """Test the installer log function with no log file."""
```
Sure.  I'll take a pass at it.

tldahlgren commented:
@alalazo I addressed your feedback and we're getting ci test and coverage results but the builds are complaining about fetching archives.  Should I try closing and re-opening?

tldahlgren commented:
> @alalazo I addressed your feedback and we're getting ci test and coverage results but the builds are complaining about fetching archives. Should I try closing and re-opening?

Well, they were there until I opened and closed the PR twice in a row. T.T

alalazo commented:
@tldahlgren The PR looks good to me. There were issues with our CI due to (I think) some change in Github apt repositories that were solved in #15458. If you can do a final update to this PR I'll merge it as soon as it goes green.

tldahlgren commented:
> @tldahlgren The PR looks good to me. There were issues with our CI due to (I think) some change in Github apt repositories that were solved in #15458. If you can do a final update to this PR I'll merge it as soon as it goes green.

Thanks @alalazo

tldahlgren commented:
@alalazo They are passing now.
